### PR TITLE
8282691: add jdb "-R" option for passing any argument to the launched debuggee process

### PR DESCRIPTION
**Pull request by plummercj**
jdb has 3 types of arguments:

1. Those that are jdb specific, such as -attach, -launch, and -listconnectors
2. Those that are passed to the JVM used to run jdb. These are all prefixed with -J, and any valid JVM argument can be passed in this manner.
3. Those that are passed to the debuggee JVM when it is launched, such as -classpath, -D<property>, and any option that starts with -X (including -XX)

The problem with the 3rd group is that (other than for -D and -X), jdb will only pass through arguments that it recognizes, and that list is very limited. When you want to launch the debuggee with an argument that jdb doesn't recognize, you have no choice but to launch the debuggee yourself (using a separate command line and using the -agentlib:jdwp argument) and then tell jdb to attach to the debuggee process. It's much easier when you can just let jdb launch the debuggee, and our nsk/jdb testing currently relies on this feature.

This PR adds the jdb -R<option> where <option> is passed through to the debuggee JVM. This allows the passing through of any JVM option, not just ones that jdb knows about. This is consistent with how jshell passes arguments to the debuggee.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8282691](https://bugs.openjdk.java.net/browse/JDK-8282691): add jdb "-R" option for passing any argument to the launched debuggee process
 * [JDK-8282830](https://bugs.openjdk.java.net/browse/JDK-8282830): add jdb "-R" option for passing any argument to the launched debuggee process (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)
 * [Kevin Walls](https://openjdk.java.net/census#kevinw) (@kevinjwalls - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7708/head:pull/7708` \
`$ git checkout pull/7708`

Update a local copy of the PR: \
`$ git checkout pull/7708` \
`$ git pull https://git.openjdk.java.net/jdk pull/7708/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7708`

View PR using the GUI difftool: \
`$ git pr show -t 7708`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7708.diff">https://git.openjdk.java.net/jdk/pull/7708.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
jshell uses -R to pass arguments through to the "remote" runtime (it's a debugger). I wonder if there is merit in doing something similar here so that jdb doesn't have to special case specific options.

**Comment by plummercj** [comment]
> jshell uses -R to pass arguments through to the "remote" runtime (it's a debugger). I wonder if there is merit in doing something similar here so that jdb doesn't have to special case specific options.

I didn't know about jshell, but I was wondering if maybe jdb should have some other prefix to use to pass args on to the debuggee. For backwards compatibility, it would have to continue to support special casing the args that it currently special cases. Maybe we push through this PR as-is to resolve this specific issue for loom and file a separate RFE to add -R support.

**Comment by plummercj** [comment]
I looked a little closer at the options that jdb does not support passing through to the debuggee (other than the two I'm adding with this PR). The more useful ones are:

- `-showversion`, which is odd because jdb supports `-version`, which is kind of useless to support because it just causes the jvm to show the version and then exit, so no chance to debug anything.
- `-verbose`
- The 4 assertion related options
- The 3 options related to native and java agents

So maybe it is worth adding `-R` at this point, although the mix of supporting some options directly, and supporting all options with `-R` is a bit ugly. You know people are going to at some point try an option that is supported and one that is not, and be wonder why one works and the other doesn't. Hopefully they look a bit closer at the help output and see they can use `-R`. Thoughts?

**Comment by AlanBateman** [comment]
> Maybe we push through this PR as-is to resolve this specific issue for loom and file a separate RFE to add -R support.

`jshell --enable-preview` allows developers to use preview features in the REPL, having `jdb --enable-preview` and passing it through to the launching debugee would be useful too.

**Comment by plummercj** [comment]
> `jshell --enable-preview` allows developers to use preview features in the REPL, having `jdb --enable-preview` and passing it through to the launching debugee would be useful too.

Is that the same as `-R--enable-preview` or `-J--enable-preview`? The help output isn't clear. It list the following options, which at first glance look like they would apply to the target process, but than that's also what `-R` is for:
```
    --class-path <path>   Specify where to find user class files
    --module-path <path>  Specify where to find application modules
    --add-modules <module>(,<module>)*
                          Specify modules to resolve, or all modules on the
                            module path if <module> is ALL-MODULE-PATHs
    --enable-native-access
                          Allow code to run restricted native methods
    --enable-preview      Allow code to depend on preview features of this release
```

**Comment by AlanBateman** [comment]
> Is that the same as `-R--enable-preview` or `-J--enable-preview`? The help output isn't clear. It list the following options, which at first glance look like they would apply to the target process, but than that's also what `-R` is for:

jshell compiles code snippets at runtime so it --enable-preview is used with the compile (along with --release or --source I assume) and also added to the options specified to the debuggee, so -R.

In the case of jdb then I think we want --enable-preview passed through to debuggees that jdb starts launching connectors, so I a bit like jshell -R.

**Comment by plummercj** [comment]
> jshell compiles code snippets at runtime so it --enable-preview is used with the compile (along with --release or --source I assume) and also added to the options specified to the debuggee, so -R.

jshell also has `-C` to pass through to the compiler, so it sounds like the list of 5 options I mentioned above (including `--enable-preview`) are treated like they were specified with both -C and -R.

> In the case of jdb then I think we want --enable-preview passed through to debuggees that jdb starts launching connectors, so I a bit like jshell -R.

Yes, that's what this PR does, in addition to passing through `--add-modules`. I'm thinking I should add `--module-path` also.

**Review by AlanBateman — APPROVED** [review]
Changes look okay, I'm just not sure if the localised resources should be updated as part of this PR as they are usually done later in the release.

**Review by kevinjwalls — APPROVED** [review]
Looks good.

Yes a "pass through unrecognised options" to the target might  be nice in future.  There are a few more module-related options (in the "java -X" help page) like --add-opens and --add-exports and a few more, which might be useful for somebody.

**Comment by plummercj** [comment]
> Changes look okay, I'm just not sure if the localised resources should be updated as part of this PR as they are usually done later in the release.

The problem is that if you don't have an entry then it throws an exception when trying to access the resource, so we include the entry but don't localize it. We had another recent case where this came up and decided then to include the unlocalized entries:

256         {"repeat is on", "Repeat is on"},
257         {"repeat is off", "Repeat is off"},
258         {"repeat usage", "Usage: repeat <on|off>"},

Note for the help output, it's all one big resource, so there is no issue with the new options not being present there (and adding unlocalized options would be messy).

**Comment by plummercj** [comment]
I found myself chasing additional options to support. I mentioned `--module-path` and then Kevin mentioned `--add-opens` and `--add-exports`. The assertion options also came to mind. So I decided to just go the `-R` route, and that's where this PR now stands. Support for `--enable-preview` and `--add-modules` have been removed.

**Comment by kevinjwalls** [comment]
Yes, that should avoid chasing future options!

There is some magic that lets us add "--add-modules", then "jdk.attach" separately, but retrieve them later as "--add-modules=jdk.attach" in the test.  I don't see what is doing that, so I ran it and verified the test works.

File: src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources.java [hunk 1]
```diff
@@ -477,8 +477,9 @@
              "    -dbgtrace [flags] print info for debugging {0}\n" +
              "    -tclient          run the application in the HotSpot(TM) Client Compiler\n" +
              "    -tserver          run the application in the HotSpot(TM) Server Compiler\n" +
+             "    -R<option>        Forward <option> to debuggee process\n" +
```
Comment (AlanBateman) [inline src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources.java:480]: The existing description start with a lowercase character so probably best to keep that consistent.

Comment (plummercj) [reply src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources.java:480]: > There is some magic that lets us add "--add-modules", then "jdk.attach" separately, but retrieve them later as "--add-modules=jdk.attach" in the test. I don't see what is doing that, so I ran it and verified the test works.

Yes, there is some magic. No, I don't know where it's done. It seems it must be done by the launcher somewhere, but I can't find it. But the hotspot arguments.cpp does expect it:

```
    } else if (match_option(option, "--add-modules=", &tail)) {
      if (!create_numbered_module_property("jdk.module.addmods", tail, addmods_count++)) {
        return JNI_ENOMEM;
      }
```
I don't see any checks for `--add-modules` without the `=`.

Comment (plummercj) [reply src/jdk.jdi/share/classes/com/sun/tools/example/debug/tty/TTYResources.java]: ok

**Comment by plummercj** [comment]
CSR [JDK-8282830](https://bugs.openjdk.java.net/browse/JDK-8282830) is ready for review.

**Comment by plummercj** [comment]
> There is some magic that lets us add "--add-modules", then "jdk.attach" separately, but retrieve them later as "--add-modules=jdk.attach" in the test. I don't see what is doing that, so I ran it and verified the test works.

Thought I had answered this yesterday, but I don't see my reply. Yes, there appears to be some magic somewhere, but I can't find it. Must be in the launcher because the hotspot arguments.cpp only accepts the "=" from of the argument: 

    } else if (match_option(option, "--add-modules=", &tail)) {
      if (!create_numbered_module_property("jdk.module.addmods", tail, addmods_count++)) {
        return JNI_ENOMEM;
      }

**Comment by plummercj** [comment]
> Thought I had answered this yesterday...

And the odd thing is I have an email with my response yesterday, but don't see it in the PR. Very strange.

**Comment by plummercj** [comment]
> > Thought I had answered this yesterday...
> 
> And the odd thing is I have an email with my response yesterday, but don't see it in the PR. Very strange.

Ah I see now. I accidentally replied in a separate conversation thread with Alan that I then marked resolved, so it was not showing up anymore.

**Review by AlanBateman — APPROVED** [review]

**Review by kevinjwalls — APPROVED** [review]

**Comment by plummercj** [comment]
Thanks Kevin and Alan!

**Comment by plummercj** [comment]
@AlanBateman and @kevinjwalls Sorry, just realized that the jdb man page needed updating. Can you review again? Thanks.

File: src/jdk.jdi/share/man/jdb.1 [hunk 2]
```diff
@@ -230,15 +230,25 @@ Runs the application in the Java HotSpot VM server.
 .RE
 .TP
 .B \f[CB]\-J\f[R]\f[I]option\f[R]
-Passes \f[I]option\f[R] to the JVM, where option is one of the options
-described on the reference page for the Java application launcher.
+Passes \f[I]option\f[R] to the JDB JVM, where option is one of the
+options described on the reference page for the Java application
+launcher.
```
Comment (plummercj) [inline src/jdk.jdi/share/man/jdb.1:235]: FYI, the change here is replacing "JVM" with "JDB JVM". The rest is just formatting.

**Review by AlanBateman — APPROVED** [review]

**Review by kevinjwalls — APPROVED** [review]

**Pull request closed without merging**